svgdigitizer/svgplot.py

```diff
@@ -28,160 +28,514 @@ class SVGPlot:
+        ...     <text x="0" y="200">x1: 0</text>
+        ...   </g>
+        ...   <g>
+        ...     <path d="M 100 200 L 100 100" />
+        ...     <text x="100" y="200">x2: 1</text>
+        ...   </g>
+        ...   <g>
+        ...     <path d="M -100 100 L 0 100" />
+        ...     <text x="-100" y="100">y1: 0</text>
+        ...   </g>
+        ...   <g>
+        ...     <path d="M -100 0 L 0 0" />
+        ...     <text x="-100" y="0">y2: 1</text>
+        ...   </g>
+        ... </svg>''')
+        >>> SVGPlot(svg).transformation
+        array([[ 0.01,  0.  ,  0.  ],
+               [ 0.  , -0.01,  1.  ],
+               [ 0.  ,  0.  ,  1.  ]])
+
+        A typical plot. Like the above but the origin is shifted and the two
+        axes are not scaled equally. Here (1000, 500) is mapped to (0, 0) and
+        (1100, 300) is mapped to (1, 1)::
+
+        >>> from io import StringIO
+        >>> svg = StringIO(r'''
+        ... <svg>
+        ...   <g>
+        ...     <path d="M 1000 600 L 1000 500" />
+        ...     <text x="1000" y="600">x1: 0</text>
+        ...   </g>
+        ...   <g>
+        ...     <path d="M 1100 600 L 1100 500" />
+        ...     <text x="1100" y="600">x2: 1</text>
+        ...   </g>
+        ...   <g>
+        ...     <path d="M 900 500 L 1000 500" />
+        ...     <text x="900" y="500">y1: 0</text>
+        ...   </g>
+        ...   <g>
+        ...     <path d="M 900 300 L 1000 300" />
+        ...     <text x="900" y="300">y2: 1</text>
+        ...   </g>
+        ... </svg>''')
+        >>> A = SVGPlot(svg).transformation
+        >>> from numpy import allclose
+        >>> allclose(A, [
+        ...   [ 0.01,  0.000, -10.00],
+        ...   [ 0.00, -0.005,   2.50],
+        ...   [ 0.00,  0.000,   1.00],
+        ... ])
+        True
+
+        A skewed plot. In this plot the axes are not orthogonal. In real plots
+        the axes might be non-orthogonal but not as much as in this
+        example. Here, one axis goes horizontally from (0, 100) to (100, 100)
+        and the other axis goes at an angle from (0, 100) to (100, 0)::
+
+        >>> from io import StringIO
+        >>> svg = StringIO(r'''
+        ... <svg>
+        ...   <g>
+        ...     <path d="M 0 200 L 0 100" />
+        ...     <text x="0" y="200">x1: 0</text>
+        ...   </g>
+        ...   <g>
+        ...     <path d="M 100 200 L 100 100" />
+        ...     <text x="100" y="200">x2: 1</text>
+        ...   </g>
+        ...   <g>
+        ...     <path d="M -100 100 L 0 100" />
+        ...     <text x="-100" y="100">y1: 0</text>
+        ...   </g>
+        ...   <g>
+        ...     <path d="M 0 0 L 100 0" />
+        ...     <text x="0" y="0">y2: 1</text>
+        ...   </g>
+        ... </svg>''')
+        >>> SVGPlot(svg, algorithm='mark-aligned').transformation
+        array([[ 0.01,  0.01, -1.  ],
+               [ 0.  , -0.01,  1.  ],
+               [ 0.  ,  0.  ,  1.  ]])
+
+        """
```

Comment:
The algorithm below is basic linear algebra. It might still be unclear how it works if you don't do these things regularly. I tried to put a lot of comments in here. If you are interested how this actually works, I am happy to talk about it next time.